### PR TITLE
contracts-bedrock: move immutables to storage for OptimismPortal

### PR DESCRIPTION
**Pull request by 0xfuturistic**
context: https://github.com/ethereum-optimism/client-pod/issues/265
security review (wip): https://www.notion.so/oplabs/DRAFT-Security-Review-Move-Immutables-to-Storage-in-L1-Contracts-56c07789fa454d8fb2df80f0b84a483b?pvs=4
legacy PR: https://github.com/ethereum-optimism/optimism/pull/6446

### CODE REVIEW

File: packages/contracts-bedrock/snapshots/storageLayout/OptimismPortal.json
Comment (tynes) [inline packages/contracts-bedrock/snapshots/storageLayout/OptimismPortal.json:78]: This is good no diff in storage layout only additions

File: packages/contracts-bedrock/src/L1/OptimismPortal.sol
Comment (tynes) [inline packages/contracts-bedrock/src/L1/OptimismPortal.sol:56]: Do you think this reads better?

File: packages/contracts-bedrock/scripts/ChainAssertions.sol
Comment (tynes) [inline packages/contracts-bedrock/scripts/ChainAssertions.sol:201]: I am confused as to why we need `_isInitialized`

Comment (0xfuturistic) [reply packages/contracts-bedrock/scripts/ChainAssertions.sol:183]: `checkOptimismMintableERC20Factory` is called at deployment of OptimismMintableERC20Factory https://github.com/ethereum-optimism/optimism/blob/9b513d7c18048bceeb132c1796cf1975fd332e91/packages/contracts-bedrock/scripts/Deploy.s.sol#L555 however, since at deployment the storage variables are set to null values now, checkOptimismMintableERC20Factory fails because the addresses don't match with dependent contracts. there are two options: removing the call to checkOptimismMintableERC20Factory from deployment i.e., removing this line: https://github.com/ethereum-optimism/optimism/blob/9b513d7c18048bceeb132c1796cf1975fd332e91/packages/contracts-bedrock/scripts/Deploy.s.sol#L555 or making checkOptimismMintableERC20Factory aware of the fact that OptimismMintableERC20Factory isn't initialized. originally, I had gone for the first approach, but this comment made me change my mind (highly recommend reading for context): https://github.com/ethereum-optimism/optimism/pull/8631#discussion_r1428267234

Comment (0xfuturistic) [reply packages/contracts-bedrock/scripts/ChainAssertions.sol:201]: `checkOptimismPortal` is called at deployment of OptimismPortal https://github.com/ethereum-optimism/optimism/blob/bb5e92e5ee49aa5099c463a391eeb33bd6308d6f/packages/contracts-bedrock/scripts/Deploy.s.sol#L503 however, since at deployment the storage variables are set to null values now, checkOptimismPortal fails because the addresses don't match with dependent contracts. there are two options: removing the call to checkOptimismPortal from deployment i.e., removing this line: https://github.com/ethereum-optimism/optimism/blob/bb5e92e5ee49aa5099c463a391eeb33bd6308d6f/packages/contracts-bedrock/scripts/Deploy.s.sol#L503 or making `checkOptimismPortal` aware of the fact that OptimismPortal isn't initialized. originally, I had gone for the first approach, but this comment made me change my mind (highly recommend reading for context): https://github.com/ethereum-optimism/optimism/pull/8631#discussion_r1428267234. thoughts?

File: packages/contracts-bedrock/src/L1/OptimismPortal.sol
Comment (0xfuturistic) [reply packages/contracts-bedrock/src/L1/OptimismPortal.sol:56]: yeah. the change aims to make the documentation consistent with the type of the variable it's for. the variable is `SuperchainConfig public superchainConfig;`, which is not an address (like the old comment says: `@notice The address of the Superchain Config contract.`) and rather is a contract (like the new comment says: `@notice Contract of the Superchain Config.`). lmk if i'm misinterpreting!

File: packages/contracts-bedrock/scripts/ChainAssertions.sol
Comment (0xfuturistic) [reply packages/contracts-bedrock/scripts/ChainAssertions.sol:201]: realized you were right 🙂

File: op-chain-ops/upgrades/l1.go
Comment (tynes) [inline op-chain-ops/upgrades/l1.go:579]: Same as other PRs, prefer to use the same values as on the chain and then check against the values in the config (if they exist) and error if mismatch

File: packages/contracts-bedrock/src/L1/OptimismPortal.sol
Comment (tynes) [inline packages/contracts-bedrock/src/L1/OptimismPortal.sol]: Do we want the same thing done here with `l2Sender` as in https://github.com/ethereum-optimism/optimism/pull/8797? cc @maurelian @mds1

File: packages/contracts-bedrock/test/L1/OptimismPortal.t.sol
Comment (tynes) [inline packages/contracts-bedrock/test/L1/OptimismPortal.t.sol]: We could use the constant `Constants.DEFAULT_L2_SENDER` here i believe instead of hardcoding the dead addr

File: packages/contracts-bedrock/test/L1/OptimismPortal.t.sol
Comment (tynes) [inline packages/contracts-bedrock/test/L1/OptimismPortal.t.sol]: These tests should be covered in `Initializable.t.sol`, can delete them if so

File: packages/contracts-bedrock/src/L1/OptimismPortal.sol
Comment (maurelian) [reply packages/contracts-bedrock/src/L1/OptimismPortal.sol]: Yes, we should.
But I will likely address that as an audit fix. 

Would it be possible to hold off on merging this PR, until the extended pause audit fixes are in. I can push to make them all ASAP to unblock this.

File: op-chain-ops/upgrades/l1.go
Comment (0xfuturistic) [reply op-chain-ops/upgrades/l1.go:579]: implemented this, but had to add `L2OutputOracle` to the `DeployConfig`, since the address wasn't there. I think this change is ok but wanted to double check with you @tynes @maurelian

File: packages/contracts-bedrock/src/L1/OptimismPortal.sol
Comment (0xfuturistic) [reply packages/contracts-bedrock/src/L1/OptimismPortal.sol]: just to confirm: no changes here in this PR?

File: packages/contracts-bedrock/test/L1/OptimismPortal.t.sol
Comment (0xfuturistic) [inline packages/contracts-bedrock/test/L1/OptimismPortal.t.sol:903]: we can delete this because we are checking for reinitialization in `Initializable.t.sol` for both the proxy and the implementation of OptimismPortal. these tests are introduced to `Initializable.t.sol` in https://github.com/ethereum-optimism/optimism/pull/8883

File: packages/contracts-bedrock/src/L1/OptimismPortal.sol
Comment (maurelian) [reply packages/contracts-bedrock/src/L1/OptimismPortal.sol]: No, the changes @tynes referenced were handled in #8797. 
Please make sure the feature branch is rebased on it, and that you're maintaining those changes.

File: op-chain-ops/upgrades/l1.go
Comment (tynes) [reply op-chain-ops/upgrades/l1.go:579]: We technically do not need to gate this behind `if config != nil` but its not the end of the world

File: packages/contracts-bedrock/src/L1/OptimismPortal.sol
Comment (tynes) [inline packages/contracts-bedrock/src/L1/OptimismPortal.sol]: You can delete line 118 since its handled on 122

**Comment by tynes** [comment]
Ready to merge after https://github.com/ethereum-optimism/optimism/pull/8629#discussion_r1448069348 is addressed

**Comment by tynes** [comment]
@0xfuturistic looks like we need to rebase and regenerate gas-snapshot

**Review by tynes — APPROVED** [review]

**Pull request merged**
